core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,170 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+#define KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Atomic.hpp>
+
+#include <functional>
+
+namespace Kokkos {
+namespace Impl {
+
+template <typename T>
+class MaybeReferenceCountedPtr {
+ protected:
+  MaybeReferenceCountedPtr(T* value_ptr)
+      : m_value_ptr(value_ptr), m_control(nullptr) {}
+
+  template <class Deleter>
+  MaybeReferenceCountedPtr(T* value_ptr, Deleter deleter)
+      : m_value_ptr(value_ptr), m_control(new Control{std::move(deleter), 1}) {}
+
+ public:
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      MaybeReferenceCountedPtr&& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+    other.m_value_ptr = nullptr;
+    other.m_control   = nullptr;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      const MaybeReferenceCountedPtr& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    if (m_control) Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      MaybeReferenceCountedPtr&& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr       = other.m_value_ptr;
+      other.m_value_ptr = nullptr;
+      m_control         = other.m_control;
+      other.m_control   = nullptr;
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      const MaybeReferenceCountedPtr& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr = other.m_value_ptr;
+      m_control   = other.m_control;
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      if (m_control) Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION ~MaybeReferenceCountedPtr() { cleanup(); }
+
+  KOKKOS_FUNCTION T* get() const noexcept { return m_value_ptr; }
+  KOKKOS_FUNCTION T& operator*() const noexcept { return *get(); }
+  KOKKOS_FUNCTION T* operator->() const noexcept { return get(); }
+
+  // checks if the stored pointer is not null
+  KOKKOS_FUNCTION explicit operator bool() const noexcept {
+    return get() != nullptr;
+  }
+
+  // checks whether the MaybeReferenceCountedPtr manages the lifetime of the
+  // object
+  KOKKOS_FUNCTION bool owner() const noexcept { return m_control != nullptr; }
```

Comment:
I would name this `is_reference_counted()` or `is_owning()` to avoid confusion with the meaning of `owner()` in, e.g., `std::owner_less<std::shared_ptr>`